controllers/extensions/addon_controller_stages.go

```diff
@@ -177,6 +178,20 @@ func (r *fetchNDeletionCheckStage) Handle(ctx context.Context) {
 	}
 	r.reqCtx.Log.V(1).Info("get addon", "generation", addon.Generation, "observedGeneration", addon.Status.ObservedGeneration)
 	r.reqCtx.UpdateCtxValue(operandValueKey, addon)
+
+	// CheckIfAddonUsedByCluster, if err, skip the deletion stage
+	if !addon.GetDeletionTimestamp().IsZero() || (addon.Spec.InstallSpec != nil && !addon.Spec.InstallSpec.GetEnabled()) {
+		used, err := CheckIfAddonUsedByCluster(ctx, r.reconciler.Client, addon.Name, r.reqCtx.Req.Namespace)
+		if err != nil {
+			r.setRequeueWithErr(err, "")
+			return
+		} else if used {
+			r.reconciler.Event(addon, corev1.EventTypeNormal, "Addon is used by some clusters",
+				"Addon is used by cluster, please check")
+			r.setRequeueAfter(time.Second, "Waiting for the cluster to end")
+			return
+		}
+	}
 	res, err := intctrlutil.HandleCRDeletion(*r.reqCtx, r.reconciler, addon, addonFinalizerName, func() (*ctrl.Result, error) {
 		r.deletionStage.Handle(ctx)
```

Comment:
Should use deleteion handle.

You can read how clusterdefintion controller solve similar problems.

Reply:
I mean the above check shall be embedded inside the deletion handler.

Reply:
I did consider this at the beginning, but there are two scenarios here: one is deleting the addon, and the other is disabling the addon. If we place the deletion judgment inside the deletion handler, then we would need to handle disabling separately. So, I combined both processes together

Reply:
I see. Make sense.